File: 백준/Silver/15649. N과 M （1）/N과 M （1）.java
import java.io.*;
import java.util.*;

public class Main {

    static int n;
    static int m;
    static int[] d = new int[10];
    static boolean[] used = new boolean[10];
    static StringBuilder sb = new StringBuilder();

    public static void backtracking(int cnt) {
        if (cnt == m) {
            for (int i = 0; i < m; i++) {
                sb.append(d[i]).append(' ');
            }
            sb.append('\n');
            return;
        }

        for (int i = 1; i <= n; i++) {
            if (used[i]) continue;
            used[i] = true;
            d[cnt] = i;
            backtracking(cnt + 1);
            used[i] = false;
        }
    }
    
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] parts = br.readLine().split(" ");
        n = Integer.parseInt(parts[0]);
        m = Integer.parseInt(parts[1]);
        backtracking(0);
        System.out.print(sb);
    }
}

File: 백준/Silver/15649. N과 M （1）/N과 M （1）.kt
import kotlin.math.*
import java.util.*

/**
 * path당 상태를 유지해야하는 것
 * - 현재 몇 개를 골랐는지: 탈출 조건
 * - 지금까지 어떤 숫자를 사용했는지 (중복 금지)
 */




fun main() = with(System.`in`.bufferedReader()) {
    val (n, m) = readLine().split(" ").map { it.toInt() }

    val used = BooleanArray(n + 1)
    val arr = IntArray(m)
    val sb = StringBuilder()
    
    fun bt(cur: Int) {
        if (cur == m) {
            arr.forEach { sb.append(it).append(' ') }
            sb.append('\n')
            return
        }

        for (i in 1..n) {
            if (used[i]) continue

            used[i] = true
            arr[cur] = i
            bt(cur + 1)
            used[i] = false
        }
    }

    bt(0)
    print(sb)
}